### ISSUE TITLE
forum.create_thread does not check category exists

### ISSUE THREAD

**Issue by mnaamani** [Joystream/joystream#3270]
TLDR: Forum runtime module: `ensure_can_create_thread` depends on `ensure_category_is_mutable()` to check category can be updated, but latter function doesn't do a check for category existance. This allows a thread to be created in a category even if that category does not exist.

https://github.com/Joystream/joystream/blob/0d14c48969d69439ac96d6021138c0b84bf50a15/runtime-modules/forum/src/lib.rs#L2214

How we found the bug..

On testing network, query-node processor failed processing a `forum.ThreadCreated` event.

![](https://user-images.githubusercontent.com/1621012/154807412-f09f0d96-322c-460e-ba53-90db6522d89a.png)

Categories are expected to start at id number 1.
Upon investigation it was found that category 0 in the chain state, but not in the query node. 
Clearly the call to `create_thread` mutated category 0 without correctly checking it exists.
I parsed the code and indeed didn't find an explicit check for the existence of the category.

`ensure_can_create_thread` depends on `ensure_category_is_mutable()` to check category can be updated.
But latter function doesn't do a check for category exists.

https://github.com/Joystream/joystream/blob/0d14c48969d69439ac96d6021138c0b84bf50a15/runtime-modules/forum/src/lib.rs#L2214

```
forum.categoryById: Category
{
  title_hash: 0x0000000000000000000000000000000000000000000000000000000000000000
  description_hash: 0x0000000000000000000000000000000000000000000000000000000000000000
  archived: false
  num_direct_subcategories: 0
  num_direct_threads: 1
  num_direct_moderators: 0
  parent_category_id: null
  sticky_thread_ids: []
}
```

Clearly the zero indicate the category was mutated

Processor Log:

```
INFO: Processing block: 0000019955-6e3a5, events count: 1 
Invalid metadata: Cannot deserialize ForumThreadMetadata! Provided bytes: (0x636f6e74657374) undefined
ERROR: Stopping the proccessor due to errors: {
  "message": "insert or update on table \"forum_thread\" violates foreign key constraint \"FK_d8cc0d0d2ff9b33ece2432c0cac\"",
  "length": 293,
  "name": "QueryFailedError",
  "severity": "ERROR",
  "code": "23503",
  "detail": "Key (category_id)=(0) is not present in table \"forum_category\".",
  "schema": "public",
  "table": "forum_thread",
  "constraint": "FK_d8cc0d0d2ff9b33ece2432c0cac",
  "file": "ri_triggers.c",
  "line": "2463",
  "routine": "ri_ReportViolation",
  "query": "INSERT INTO \"forum_thread\"(\"id\", \"created_at\", \"created_by_id\", \"updated_at\", \"updated_by_id\", \"deleted_at\", \"deleted_by_id\", \"version\", \"author_id\", \"category_id\", \"title\", \"initial_post_id\", \"visible_posts_count\", \"is_sticky\", \"status\", \"is_visible\") VALUES ($1, $2, $3, $4, DEFAULT, DEFAULT, DEFAULT, $5, $6, $7, $8, DEFAULT, $9, $10, $11, $12) RETURNING \"created_at\", \"updated_at\", \"version\"",
  "parameters": [
    "1",
    "2022-02-19T06:39:36.000Z",
    "81PbMapRf",
    "2022-02-19T06:39:36.000Z",
    1,
    "100",
    "0",
    "contest",
    1,
    false,
    "{\"isTypeOf\":\"ThreadStatusActive\"}",
    true
  ]
} name: QueryFailedError, message: insert or update on table "forum_thread" violates foreign key constraint "FK_d8cc0d0d2ff9b33ece2432c0cac"
```

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201709380879215/1201870684875007) by [Unito](https://www.unito.io)

**Comment by mnaamani** [comment]
For background the thread was created through creation of the bounty.
When the forum bug is fixed we should also insure the the bounty creation + thread creation is "atomic".
Also worth asking, how does the bounty module select what category to use?

**Comment by bedeho** [comment]
> to check category can be updated, but latter function doesn't do a check for category existance

This is concerning, great find.

> Also worth asking, how does the bounty module select what category to use?

Excellent question

**Comment by mnaamani** [comment]
> > Also worth asking, how does the bounty module select what category to use?
> 
> Excellent question

I checked the runtime code. I don't see the bounty runtime module actually creating the thread. I think it is Pioneer itself, that is first creating a thread, before creating the bounty. So it is not as concerning, only the forum bug needs to be addressed.

Can you confirm this @thesan ?

**Comment by mnaamani** [comment]
Fixed in https://github.com/Joystream/joystream/pull/3279
